rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java

```diff
@@ -0,0 +1,153 @@
+ * ByteBuf.
+ */
+public class CollectBytes implements Transformer<ByteBuf, ByteBuf> {
+
+    private static final CollectBytes INSTANCE = new CollectBytes();
+
+    /**
+     * Collect all emitted ByteBufs into a single ByteBuf.
+     */
+    public static Transformer<ByteBuf, ByteBuf> all() {
+        return INSTANCE;
+    }
+
+    /**
+     * Collect all emitted ByteBufs into a single ByteBuf until maxBytes have
+     * been collected. If more than maxBytes are received this unsubscribe from
+     * the upstream Observable and will emit a
+     * {@link TooMuchDataException}. {@link TooMuchDataException#getCause()}
+     * will contain an
+     * {@link OnErrorThrowable.OnNextValue} with the bytes accumulated before the exception
+     * was thrown.
+     */
+    public static Transformer<ByteBuf, ByteBuf> upTo(final int maxBytes) {
+        return new Transformer<ByteBuf, ByteBuf>() {
+            @Override
+            public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+                return upstream
+                        .compose(new TransformerLimitByteBufTotalSize(maxBytes))
+                        .compose(INSTANCE)
+                        .flatMap(new Func1<ByteBuf, Observable<ByteBuf>>() {
+                            @Override
+                            public Observable<ByteBuf> call(ByteBuf buf) {
+                                return buf.readableBytes() > maxBytes ?
+                                        Observable.<ByteBuf>error(OnErrorThrowable.addValueAsLastCause(new TooMuchDataException(), buf)) :
+                                        Observable.just(buf);
+                            }
+                        });
+            }
+        };
+    }
+
+    private CollectBytes() {
+        // do nothing
+    }
+
+    @Override
+    public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+        return upstream.collect(
+                new Func0<CompositeByteBuf>() {
+                    @Override
+                    public CompositeByteBuf call() {
+                        return Unpooled.compositeBuffer();
+                    }
+                },
+                new Action2<CompositeByteBuf, ByteBuf>() {
+                    @Override
+                    public void call(CompositeByteBuf collector, ByteBuf buf) {
+                        int i = buf.readableBytes();
+                        collector.addComponent(buf);
+                        collector.writerIndex(collector.writerIndex() + i);
+                    }
+                }
+            )
+            .cast(ByteBuf.class);
+    }
+
+    public static class TransformerLimitByteBufTotalSize implements Observable.Transformer<ByteBuf, ByteBuf> {
+
+        private final int maxBytes;
+
+        public TransformerLimitByteBufTotalSize(int maxBytes) {
+            this.maxBytes = maxBytes;
+        }
+
+        @Override
+        public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+            return upstream.scan(
```

Comment:
>> Also added a comment about (not) using autorelease since the whole thing dies if you do.

Hmm ... that is unfortunate 😞 but I don't think there is a way to avoid this easily.